crates/subspace-farmer/src/plotting.rs

```diff
@@ -263,19 +260,30 @@ async fn background_plotting<T: RpcClient + Clone + Send + 'static>(
                         }
                         if let Some(plot) = weak_plot.upgrade() {
                             let pieces = Arc::new(pieces);
-                            // TODO: There is no internal mapping between pieces and their indexes yet
-                            // TODO: Then we might want to send indexes as a separate vector
-                            if let Err(error) =
-                                plot.write_many(Arc::clone(&pieces), piece_index_offset)
-                            {
-                                error!("Failed to write encoded pieces: {}", error);
-                            }
-                            if let Err(error) = farmer_data
-                                .commitments
-                                .create_for_pieces(&pieces, piece_index_offset)
-                            {
-                                error!("Failed to create commitments for pieces: {}", error);
+
+                            match plot.write_many(Arc::clone(&pieces), piece_index_offset) {
+                                Ok((offsets, old_pieces)) => {
+                                    if let Err(error) =
+                                        farmer_data.commitments.remove_pieces(old_pieces)
+                                    {
+                                        error!(
+                                            "Failed to remove old commitments for pieces: {}",
+                                            error
+                                        );
+                                    }
+
+                                    if let Err(error) =
+                                        farmer_data.commitments.create_for_pieces(&pieces, offsets)
```

Comment:
It might happen that pieces that were attempted to be written, but ended up not actually skipped because they exceeded plot size and were further from node ID than the furthest present piece. Do I understand correctly that this will not work correctly in that case and will still create commitments for those?

Reply:
Yes, you are actually right. Should I make a regression for this case?

Reply:
What do you mean by regression?

Reply:
Fixed and added `TODO` for this test for now. It would require to add background plotting tests which (I think) would require reworking plotting a bit and making it more modular.